### PR TITLE
Bug fix/add aql collection count cache

### PR DESCRIPTION
**Pull request by jsteemann**
Cache the number of documents in a collection for execution plans, in order to reduce the number of HTTP requests in the cluster.
There is also an accompanying enterprise PR: https://github.com/arangodb/enterprise/pull/132

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/755/

File: arangod/Transaction/CountCache.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      // not yet expired
+      return count;
+    }
+  }
+  return CountCache::NotPopulated;
+}
+
+void CountCache::store(int64_t value) {
+  TRI_ASSERT(value >= 0);
+  timestamp.store(TRI_microtime(), std::memory_order_acquire);
```
Comment (mpoeter) [inline arangod/Transaction/CountCache.cpp]: A store can never have acquire semantics (some implementations like the MSVC STL even have according assertions in debug builds).
The operations on `count` already have the correct memory orders to establish the required happens-before relation, so all operations on `timestamp` (including the load in `get`) can be relaxed.

Comment (jsteemann) [reply arangod/Transaction/CountCache.cpp]: @mpoeter : oh yes, makes a lot of sense. Thanks for pointing me at it!

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/757/

**Comment by mpoeter** [comment]
I just realized that there is a race condition since `timestamp` is stored before `count`. Assuming that the cache can be updated multiple times (i.e., `count != NotPopulated`): The thread in `store` could have updated `timestamp`, but not yet `count`. Another thread could see the new `timestamp` value, conclude that it is up to date and return the old, outdated `count` value.

I would therefore suggest to change the implementation as follows:
```C++
int64_t CountCache::get() const {
  return count.load(std::memory_order_relaxed);
}

int64_t CountCache::get(double ttl) const {
  // (1) - this acquire-load synchronizes with the release-store(2)
  double ts = timestamp.load(std::memory_order_acquire);
  if (ts + Ttl > TRI_microtime()) {
    // not yet expired
    return get();
  }
  return CountCache::NotPopulated;
}

void CountCache::store(int64_t value) {
  TRI_ASSERT(value >= 0);
  count.store(value, std::memory_order_relaxed);
  // (2) - this release-store synchronizes with the acquire-load (1)
  timestamp.store(TRI_microtime(), std::memory_order_release);
}
```

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/760/

File: arangod/VocBase/LogicalCollection.h [hunk 2]
```diff
@@ -338,6 +340,8 @@ class LogicalCollection: public LogicalDataSource {
   // Get a reference to this KeyGenerator.
   // Caller is not allowed to free it.
   inline KeyGenerator* keyGenerator() const { return _keyGenerator.get(); }
+  
+  transaction::CountCache& countCache() { return _countCache; }
```
Comment (graetzer) [inline arangod/VocBase/LogicalCollection.h]: can't we move this into the CoordinatorCollection ? The LogicalCollection has too many members that are only used half the time

**Comment by jsteemann** [comment]
It is intentionally in LogicalCollection, as the cache is used for local collections too, not just on the coordinator.

**Pull request merged**